In [3]:
import os
import cv2
import numpy as np
from tensorflow.keras import utils
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten, MaxPool2D, Conv2D
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from tqdm import tqdm


IMPORT DATA:

Import the data set from the location where the file is downloaded

In [6]:
Data_Dir = 'F:\study\sem 10\cvpr\VGG 16\dataset'
TEST_DATA_DIR = os.path.join(Data_Dir, 'asl_alphabet_test/asl_alphabet_test')
TRAIN_DATA_DIR = os.path.join(Data_Dir, 'asl_alphabet_train/asl_alphabet_train')

EXTRACT Train Data:

Extract all the train data in CATEGORIES variable

In [5]:
CATEGORIES = []

for i in os.listdir(TRAIN_DATA_DIR):
    CATEGORIES.append(i)
    
print(CATEGORIES)

['A', 'B', 'C', 'D', 'del', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'nothing', 'O', 'P', 'Q', 'R', 'S', 'space', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']


PRE PROCESS DATA:

Pre prosessing of all data is done here

In [17]:
images = []
labels = []
img_size = 32,32

for c in CATEGORIES:
    path = os.path.join(TRAIN_DATA_DIR, c) # 'E:/DATASETS/classification/Face Mask Dataset/train/WithMask'
    class_num = CATEGORIES.index(c) # 0
    for img in tqdm(os.listdir(path)):
        try:
            img_array = cv2.imread(os.path.join(path, img))   # read the image
            img_resized = cv2.resize(img_array, img_size)  # resize the image
            images.append(img_resized) # [ [img, 0], [], [], [], ...., []]
            labels.append(class_num)
        except WException as e:
            pass
print(labels.shape)

100%|█████████████████████████████████████████████████████████████████████████████| 3000/3000 [00:03<00:00, 864.25it/s]


AttributeError: 'list' object has no attribute 'shape'

In [16]:
images = np.array(images)
images = images.astype('float32')/255.0
labels = utils.to_categorical(labels)
x_train, x_test, y_train, y_test = train_test_split(images, labels, test_size = 0.1)
print(labels.shape)

(87000, 29, 2, 2)


VGG 16 Architecture:

In [11]:
model = keras.Sequential([
    keras.Input(shape=(32,32,3)),
    layers.Conv2D(filters=64, kernel_size=(3,3), padding="same", activation="relu"),
    layers.Conv2D(filters=64, kernel_size=(3,3), padding="same", activation="relu"),
    layers.MaxPooling2D(pool_size=(2,2), strides=(2,2)),
    
    layers.Conv2D(filters=128, kernel_size=(3,3), padding="same", activation="relu"),
    layers.Conv2D(filters=128, kernel_size=(3,3), padding="same", activation="relu"),
    layers.MaxPooling2D(pool_size=(2,2), strides=(2,2)),
    
    layers.Conv2D(filters=256, kernel_size=(3,3), padding="same", activation="relu"),
    layers.Conv2D(filters=256, kernel_size=(3,3), padding="same", activation="relu"),
    layers.Conv2D(filters=256, kernel_size=(3,3), padding="same", activation="relu"),
    layers.MaxPooling2D(pool_size=(2,2), strides=(2,2)),
    
    layers.Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"),
    layers.Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"),
    layers.Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"),
    layers.MaxPooling2D(pool_size=(2,2), strides=(2,2)),
    
    layers.Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"),
    layers.Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"),
    layers.Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"),
    layers.MaxPooling2D(pool_size=(2,2), strides=(2,2)),
    
    layers.Flatten(),
    layers.Dense(4096, activation='relu'),
    layers.Dense(4096, activation='relu'),
    layers.Dense(29, activation='softmax')
])


Compile the model and set the learnimg rate here and show the model summary and after that train the model

In [13]:
learning_rate = 0.0001
adam = Adam(lr=learning_rate)
model.compile(optimizer=adam, loss='categorical_crossentropy', metrics=['accuracy'])
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_13 (Conv2D)           (None, 32, 32, 64)        1792      
_________________________________________________________________
conv2d_14 (Conv2D)           (None, 32, 32, 64)        36928     
_________________________________________________________________
max_pooling2d_5 (MaxPooling2 (None, 16, 16, 64)        0         
_________________________________________________________________
conv2d_15 (Conv2D)           (None, 16, 16, 128)       73856     
_________________________________________________________________
conv2d_16 (Conv2D)           (None, 16, 16, 128)       147584    
_________________________________________________________________
max_pooling2d_6 (MaxPooling2 (None, 8, 8, 128)         0         
_________________________________________________________________
conv2d_17 (Conv2D)           (None, 8, 8, 256)        

In [14]:
history = model.fit(x_train, y_train, batch_size=128, epochs=5, validation_split=0.1)

Epoch 1/5
551/551 [==============================] - 51s 89ms/step - loss: 2.8956 - accuracy: 0.1133 - val_loss: 1.3202 - val_accuracy: 0.5193
Epoch 2/5
551/551 [==============================] - 48s 87ms/step - loss: 1.0047 - accuracy: 0.6318 - val_loss: 0.6224 - val_accuracy: 0.7700
Epoch 3/5
551/551 [==============================] - 36s 65ms/step - loss: 0.3762 - accuracy: 0.8635 - val_loss: 0.3328 - val_accuracy: 0.8796
Epoch 4/5
551/551 [==============================] - 24s 43ms/step - loss: 0.1800 - accuracy: 0.9384 - val_loss: 0.0832 - val_accuracy: 0.9736
Epoch 5/5
551/551 [==============================] - 24s 44ms/step - loss: 0.1290 - accuracy: 0.9560 - val_loss: 0.1746 - val_accuracy: 0.9427


As you can see we done the training seisson here and reached the 95% of training accuracy and 94% of the validation accuracy which look pretty decent and good model 